### DISCUSSION TITLE
On/Off Switch using uiComponent

### DISCUSSION THREAD

**Discussion by dnaldoog** [RomanKubiak/ctrlr#274] [Show and tell]
Here is a basic on/off switch using uiComponent and panel:getGlobalVariable() and bit.bxor(n,1) for toggling between 0 and 1.

[on off switch_1_2.zip](https://github.com/RomanKubiak/ctrlr/files/6409625/on.off.switch_1_2.zip)

![chorusonoffuicomponent](https://user-images.githubusercontent.com/76831683/116784016-d47dda00-aacc-11eb-8be4-364b5b2b0cd1.png)

**Comment by Godlike-Productions** [comment]
Cool. I'm sure I'll need this before too long.

I've just had to do a button group with Image Buttons (selecting 1 button will unselect the the others WITHOUT sending value change - otherwise it get's stuck in the mother of loops)

I'll post the code up shortly.

**Reply by dnaldoog** [reply]
Here is code I use for radio buttons using source to avoid infinite loops.

**Table t{}** contains the names of the buttons in the radio group:
<pre><code>
ProgramSelect = function(mod,value,source)
    if source == 4 then
        local sName = L(mod:getName())
        local t = {"PROGRAM01", "PROGRAM02", "PROGRAM03", "PROGRAM04"} -- local table of 4 radio buttons
        for i, v in pairs(t) do
            if sName == v then
                local sysMess = string.format("F0 00 00 23 01 43 C1 00 %.2X F7", i - 1)
                panel:sendMidiMessageNow(CtrlrMidiMessage(sysMess))
            else
                _G[v]:setValue(0, true) 
               -- don't use panel:getComponent(v) // use panel:getModulatorByName(v)
            end
        end
    end
end -- source == 4 (user clicked on a button - not lua generated - that would be 5 or 6) end

--[[ 
*NOTE: Where PROGRAM01 etc are initialised in an init script
"Called when the panel has finished loading"
PROGRAM01=panel:getModulatorByName("PROGRAM01")
PROGRAM02=panel:getModulatorByName("PROGRAM02")
PROGRAM03=panel:getModulatorByName("PROGRAM03")
PROGRAM04=panel:getModulatorByName("PROGRAM04")
--]]

</code>
</pre>

https://ctrlr.org/radio-button-code-generator/

**Reply by Godlike-Productions** [reply]
Oh nice, that's probably an even nicer way of doing it.

Does source 4 work if a button is pressed via MIDI (ie a user has mapped an Ableton Modulator to press the button)?

**Reply by dnaldoog** [reply]
https://github.com/RomanKubiak/ctrlr/discussions/200

I think in this post Goodweather talks about <code>source=2</code> being passed in for incoming. 

Maybe set up some console() code to detect what you get then do something like <code>if source == 4 or source ==2 then … end</code> perhaps?.

**Reply by damiensellier** [reply]
I used the radio button switch trick today and the source = 4 on mouse down (GUI source )and when LUA starts to loop source = 5 .
So the statement if source = 4 is mandatory to avoid breaking the time-space-continuum and destroying the universe.
 Initial values are required on startup as well.

and as a recap here are the source values : 

```
value(“initialValue”, 0),
value(“changedByHost”, 1),
value(“changedByMidiIn”, 2),
value(“changedByMidiController”, 3),
value(“changedByGUI”, 4),
value(“changedByLua”, 5),
value(“changedByProgram”, 6),
value(“changedByLink”, 7),
value(“changeByUnknown”, 8)
```
